packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_prop/component_ref_prop.rs

```diff
@@ -0,0 +1,105 @@
+use std::rc::Rc;
+
+use crate::{
+    components::prelude::*,
+    props::ContentFilter,
+    state::types::{component_refs::ComponentRef, content_refs::ContentRef},
+};
+
+/// A prop that references a single component
+#[derive(Debug)]
+pub struct ComponentRefProp {
+    /// The data query that indicates how the dependencies of this prop will be created.
+    data_query: DataQuery,
+
+    component_to_select: Option<ComponentToSelect>,
+}
+
+#[derive(Debug)]
+enum ComponentToSelect {
+    First,
+    Last,
+}
+
+impl ComponentRefProp {
+    /// Creates a ComponentRefs prop that returns the last child with component_type
+    pub fn new_from_last_matching_child(component_type: &'static str) -> Self {
+        ComponentRefProp {
+            data_query: DataQuery::ComponentRefs {
+                container: PropSource::Me,
+                filter: Rc::new(ContentFilter::IsType(component_type)),
+            },
+            component_to_select: Some(ComponentToSelect::Last),
+        }
+    }
+
+    /// Creates a ComponentRefs prop that returns the first child with component_type
+    pub fn new_from_first_matching_child(component_type: &'static str) -> Self {
+        ComponentRefProp {
+            data_query: DataQuery::ComponentRefs {
+                container: PropSource::Me,
+                filter: Rc::new(ContentFilter::IsType(component_type)),
+            },
+            component_to_select: Some(ComponentToSelect::First),
+        }
+    }
+
+    pub fn new_self_ref() -> Self {
+        ComponentRefProp {
+            data_query: DataQuery::SelfRef,
+            component_to_select: None,
+        }
+    }
+}
+
+/// Structure to hold data generated from the data queries
+#[derive(TryFromDataQueryResults, Debug)]
+#[data_query(query_trait = DataQueries, pass_data = &ComponentRefProp)]
+struct RequiredData {
+    refs: PropView<prop_type::ContentRefs>,
+}
+impl DataQueries for RequiredData {
+    fn refs_query(arg: &ComponentRefProp) -> DataQuery {
+        arg.data_query.clone()
+    }
+}
+
+impl PropUpdater for ComponentRefProp {
+    type PropType = prop_type::ComponentRef;
+
+    fn data_queries(&self) -> Vec<DataQuery> {
+        vec![self.data_query.clone()]
```

Comment:
Here's a suggested patch. (Due to a known issue, I apparently catch attach a .patch file when using Linux, so I pasted here.)

```
diff --git a/packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_prop/component_ref_prop.rs b/packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_prop/component_ref_prop.rs
index c35ec9cf7..cafd8eeb7 100644
--- a/packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_prop/component_ref_prop.rs
+++ b/packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_prop/component_ref_prop.rs
@@ -54,13 +54,13 @@ impl ComponentRefProp {
 
 /// Structure to hold data generated from the data queries
 #[derive(TryFromDataQueryResults, Debug)]
-#[data_query(query_trait = DataQueries, pass_data = &ComponentRefProp)]
+#[data_query(query_trait = DataQueries, pass_data = &DataQuery)]
 struct RequiredData {
     refs: PropView<prop_type::ContentRefs>,
 }
 impl DataQueries for RequiredData {
-    fn refs_query(arg: &ComponentRefProp) -> DataQuery {
-        arg.data_query.clone()
+    fn refs_query(query: &DataQuery) -> DataQuery {
+        query.clone()
     }
 }
 
@@ -68,7 +68,7 @@ impl PropUpdater for ComponentRefProp {
     type PropType = prop_type::ComponentRef;
 
     fn data_queries(&self) -> Vec<DataQuery> {
-        vec![self.data_query.clone()]
+        RequiredData::data_queries_vec(&self.data_query)
     }
 
     fn calculate(&self, data: DataQueryResults) -> PropCalcResult<Self::PropType> {
```